### PR TITLE
Discrete interpolation matrix

### PR DESCRIPTION
**Pull request by jorgensd**
Creation of interpolation matrix A  from space V->W such that
u \in V, w in W such that  Au = w.

Not tested with AMS solvers yet, as pybindings are missing.

### CODE REVIEW

File: cpp/dolfinx/fem/discreteoperators.h [hunk 1]
```diff
@@ -123,4 +123,199 @@ void assemble_discrete_gradient(const fem::FunctionSpace& V0,
     mat_set(dofmap1->cell_dofs(c), dofmap0->cell_dofs(c), Ae);
   }
 }
+
+/// @brief Assemble an interpolation operator matrix
+///
+/// This function creates a discrete operator \f$A\f$ that interpolates
+/// a function \f$V_0\f$ into a compatible space \f$V_1\f$, i.e.
```
Comment (garth-wells) [inline cpp/dolfinx/fem/discreteoperators.h:130]: What is a 'compatible space'?

File: cpp/dolfinx/fem/discreteoperators.h [hunk 2]
```diff
@@ -123,4 +123,199 @@ void assemble_discrete_gradient(const fem::FunctionSpace& V0,
...
+  const std::size_t dim0 = element0->space_dimension() / bs0;
+  const std::size_t value_size_ref0 = element0->reference_value_size() / bs0;
+  const std::size_t value_size0 = element0->value_size() / bs0;
+  const std::size_t value_size1 = element1->value_size() / bs1;
+
+  // Get geometry data
+  const fem::CoordinateElement& cmap = mesh->geometry().cmap();
+  const graph::AdjacencyList<std::int32_t>& x_dofmap
+      = mesh->geometry().dofmap();
+  // FIXME: Add proper interface for num coordinate dofs
```
Comment (garth-wells) [inline cpp/dolfinx/fem/discreteoperators.h:191]: Let's stop reproducing this everywhere.

File: cpp/dolfinx/fem/discreteoperators.h [hunk 3]
```diff
@@ -123,4 +123,199 @@ void assemble_discrete_gradient(const fem::FunctionSpace& V0,
...
+  const std::size_t num_dofs_g = x_dofmap.num_links(0);
+  xtl::span<const double> x_g = mesh->geometry().x();
+
+  // Evaluate coordinate map basis at reference interpolation points
+  xt::xtensor<double, 4> phi(cmap.tabulate_shape(1, X.shape(0)));
+  xt::xtensor<double, 2> dphi;
+  cmap.tabulate(1, X, phi);
+  dphi = xt::view(phi, xt::range(1, tdim + 1), 0, xt::all(), 0);
+
+  // Evaluate v basis functions at reference interpolation points
```
Comment (garth-wells) [inline cpp/dolfinx/fem/discreteoperators.h:201]: Basis functions belong to the function space. Which space? `v` isn't a parameter.

File: cpp/dolfinx/fem/discreteoperators.h [hunk 4]
```diff
@@ -123,4 +123,199 @@ void assemble_discrete_gradient(const fem::FunctionSpace& V0,
...
+
+  // Create working arrays
+  xt::xtensor<double, 3> basis0({X.shape(0), dim0, value_size0});
+  xt::xtensor<double, 3> basis_reference0({X.shape(0), dim0, value_size_ref0});
+  xt::xtensor<double, 2> coordinate_dofs({num_dofs_g, gdim});
+  xt::xtensor<double, 3> J({X.shape(0), gdim, tdim});
+  xt::xtensor<double, 3> K({X.shape(0), tdim, gdim});
+  std::vector<double> detJ(X.shape(0));
+
+  // Get interpolation operator
```
Comment (garth-wells) [inline cpp/dolfinx/fem/discreteoperators.h:214]: Which interpolation operator, i.e. V0 -> V1, something else?

File: cpp/dolfinx/fem/discreteoperators.h [hunk 5]
```diff
@@ -123,4 +123,199 @@ void assemble_discrete_gradient(const fem::FunctionSpace& V0,
...
+/// @brief Assemble an interpolation operator matrix
+///
+/// This function creates a discrete operator \f$A\f$ that interpolates
+/// a function \f$V_0\f$ into a compatible space \f$V_1\f$, i.e.
+/// \f$V_0 \rightarrow V_1\f$. If \f$u_0\f$ is the
+/// degree-of-freedom vector associated with \f$V_0\f$, then
+/// \f$u_1=Au_0\f$ where \f$u_1\f$ is the degrees-of-freedom vector for
+/// interpolating function in the \f$V_1\f$ space.
+///
+/// @warning This function relies on the user supplying appropriate
```
Comment (garth-wells) [inline cpp/dolfinx/fem/discreteoperators.h:136]: The parameter descriptions don't add anything on this.

File: cpp/dolfinx/fem/discreteoperators.h [hunk 6]
```diff
@@ -123,4 +123,199 @@ void assemble_discrete_gradient(const fem::FunctionSpace& V0,
...
+    // Pull back the physical values to the reference of output space
+    for (std::size_t p = 0; p < basis_values.shape(0); ++p)
+    {
+      auto _K = xt::view(K, p, xt::all(), xt::all());
+      auto _J = xt::view(J, p, xt::all(), xt::all());
+      auto _u = xt::view(basis_values, p, xt::all(), xt::all());
+      auto _U = xt::view(mapped_values, p, xt::all(), xt::all());
+      pull_back_fn1(_U, _u, _K, 1.0 / detJ[p], _J);
+    }
+    // Apply interpolation matrix
```
Comment (garth-wells) [inline cpp/dolfinx/fem/discreteoperators.h:307]: Apply to what?

**Review by garth-wells — CHANGES_REQUESTED** [review]

File: cpp/dolfinx/fem/discreteoperators.h (hunk 1 above)
Comment (jorgensd) [reply cpp/dolfinx/fem/discreteoperators.h:130]: A space with the correct value shape. I.e. you cannot interpolate a scalar function into a vector space or opposite.

File: cpp/dolfinx/fem/discreteoperators.h (hunk 2 above)
Comment (jorgensd) [inline cpp/dolfinx/fem/discreteoperators.h:191]:
```suggestion
```

File: cpp/dolfinx/fem/discreteoperators.h (hunk 3 above)
Comment (jorgensd) [inline cpp/dolfinx/fem/discreteoperators.h:200]:
```suggestion
  // Evaluate V0 basis functions at reference interpolation points for V1
```

File: cpp/dolfinx/fem/discreteoperators.h [hunk 7]
```diff
@@ -123,4 +123,198 @@ void assemble_discrete_gradient(const fem::FunctionSpace& V0,
...
+/// This function creates a discrete operator \f$A\f$ that interpolates
+/// a function \f$V_0\f$ into a compatible space \f$V_1\f$, i.e.
+/// \f$V_0 \rightarrow V_1\f$. If \f$u_0\f$ is the
+/// degree-of-freedom vector associated with \f$V_0\f$, then
+/// \f$u_1=Au_0\f$ where \f$u_1\f$ is the degrees-of-freedom vector for
+/// interpolating function in the \f$V_1\f$ space.
+///
+/// @warning This function relies on the user supplying appropriate
+/// input and output spaces. See parameter descriptions.
+///
```
Comment (jorgensd) [inline cpp/dolfinx/fem/discreteoperators.h:138]:
```suggestion
```

File: cpp/dolfinx/fem/discreteoperators.h [hunk 8]
```diff
@@ -123,4 +123,195 @@ void assemble_discrete_gradient(const fem::FunctionSpace& V0,
...
+
+  using u1_t = xt::xview<decltype(basis_values)&, std::size_t,
+                         xt::xall<std::size_t>, xt::xall<std::size_t>>;
+  using U1_t = xt::xview<decltype(mapped_values)&, std::size_t,
+                         xt::xall<std::size_t>, xt::xall<std::size_t>>;
+  auto pull_back_fn1 = element1->map_fn<U1_t, u1_t, K_t, J_t>();
+
+  std::vector<T> A(element1->space_dimension() * element0->space_dimension());
+  std::vector<T> local1(element1->space_dimension());
+  auto _A = xt::adapt(
```
Comment (IgorBaratta) [inline cpp/dolfinx/fem/discreteoperators.h:238]: Unused variable:
```suggestion

```

File: cpp/dolfinx/fem/discreteoperators.h (hunk 4 above)
Comment (garth-wells) [reply cpp/dolfinx/fem/discreteoperators.h:210]: Expand the docstring - doesn't mean much to the reader as it is.

File: cpp/dolfinx/fem/discreteoperators.h [hunk 9]
```diff
@@ -123,4 +123,195 @@ void assemble_discrete_gradient(const fem::FunctionSpace& V0,
...
+
+  // Create working arrays
+  xt::xtensor<double, 3> basis0({X.shape(0), dim0, value_size0});
+  xt::xtensor<double, 3> basis_reference0({X.shape(0), dim0, value_size_ref0});
+  xt::xtensor<double, 2> coordinate_dofs({num_dofs_g, gdim});
+  xt::xtensor<double, 3> J({X.shape(0), gdim, tdim});
+  xt::xtensor<double, 3> K({X.shape(0), tdim, gdim});
+  std::vector<double> detJ(X.shape(0));
+
+  // Get interpolation operator
```
Comment (jorgensd) [inline cpp/dolfinx/fem/discreteoperators.h:210]:
```suggestion
  /// Get the interpolation operator (matrix) `Pi` that maps a function
  /// evaluated at the interpolation points to the element degrees of
  /// freedom, i.e. dofs = Pi f_x
```

File: cpp/dolfinx/fem/discreteoperators.h [hunk 10]
```diff
@@ -123,4 +123,196 @@ void assemble_discrete_gradient(const fem::FunctionSpace& V0,
     mat_set(dofmap1->cell_dofs(c), dofmap0->cell_dofs(c), Ae);
   }
 }
+
+/// @brief Assemble an interpolation operator matrix
+///
+/// This function creates a discrete operator \f$A\f$ that interpolates
+/// a function \f$V_0\f$ into a compatible space \f$V_1\f$, i.e.
```
Comment (jorgensd) [inline cpp/dolfinx/fem/discreteoperators.h:130]:
```suggestion
/// a function \f$V_0\f$ into a compatible space (a space with the same value size) \f$V_1\f$, i.e.
```

File: cpp/dolfinx/fem/discreteoperators.h [hunk 11]
```diff
@@ -123,4 +123,196 @@ void assemble_discrete_gradient(const fem::FunctionSpace& V0,
     mat_set(dofmap1->cell_dofs(c), dofmap0->cell_dofs(c), Ae);
   }
 }
+
+/// @brief Assemble an interpolation operator matrix
+///
+/// This function creates a discrete operator \f$A\f$ that interpolates
+/// a function \f$V_0\f$ into a compatible space (a space with the same value size) \f$V_1\f$, i.e.
+/// \f$V_0 \rightarrow V_1\f$. If \f$u_0\f$ is the
```
Comment (jorgensd) [inline cpp/dolfinx/fem/discreteoperators.h:131]:
```suggestion
/// a function \f$V_0\f$ into a compatible space (same value shape)
///  \f$V_1\f$, i.e. \f$V_0 \rightarrow V_1\f$. If \f$u_0\f$ is the
```

File: cpp/dolfinx/fem/discreteoperators.h [hunk 12]
```diff
@@ -123,4 +123,196 @@ void assemble_discrete_gradient(const fem::FunctionSpace& V0,
...
+
+  using u1_t = xt::xview<decltype(basis_values)&, std::size_t,
+                         xt::xall<std::size_t>, xt::xall<std::size_t>>;
+  using U1_t = xt::xview<decltype(mapped_values)&, std::size_t,
+                         xt::xall<std::size_t>, xt::xall<std::size_t>>;
+  auto pull_back_fn1 = element1->map_fn<U1_t, u1_t, K_t, J_t>();
+
+  std::vector<T> A(element1->space_dimension() * element0->space_dimension());
+  std::vector<T> local1(element1->space_dimension());
+      A, {element1->space_dimension(), element0->space_dimension()});
```
Comment (jorgensd) [inline cpp/dolfinx/fem/discreteoperators.h:240]:
```suggestion
```

File: cpp/dolfinx/fem/discreteoperators.h (hunk 2 above)
Comment (garth-wells) [reply cpp/dolfinx/fem/discreteoperators.h]: I meant stop reproducing the bad code, not stop reproducing the message that the code is bad.

Comment (jorgensd) [reply cpp/dolfinx/fem/discreteoperators.h]: Do you have a suggestion as to how you would like to see this resolved? As we currently use a constant width adjacency list for the dof map:
```
  const std::size_t num_dofs_g = x_dofmap.num_links(0);
 ```  
this gives us the number of dofs per cell.
If we want to make the code executable for P-refinement, we would need to move the definition of num dofs g and coordinate_dofs inside the cell loop.

Comment (garth-wells) [reply cpp/dolfinx/fem/discreteoperators.h]: > Do you have a suggestion as to how you would like to see this resolved? As we currently use a constant width adjacency list for the dof map:
> 
> ```
>   const std::size_t num_dofs_g = x_dofmap.num_links(0);
> ```
> 
> this gives us the number of dofs per cell. If we want to make the code executable for mixed-topology meshes we would need to move the definition of num dofs g and coordinate_dofs inside the cell loop.

The dimension should probablt come from the `CoordinateElement`. ` x_dofmap.num_links(0)` takes the size from the first entry, but that requires the data to be non-empty.

**Pull request merged**